### ISSUE TITLE
KeyVault swagger spec does not call out 204 in their DELETE responses

### ISSUE THREAD

**Issue by iouri-s** [Azure/azure-rest-api-specs#921]
As a result, 204s result in CloudException being thrown with no Body. Most RP SDKs do not throw on 204, and KeyVault should act consistently

**Comment by vishrutshah** [comment]
@pomortaz Could you please look into this. It'd be great if we have defined `204` status codes in responses so than AutoRest would not throw. 

Reference: https://github.com/Azure/azure-rest-api-specs/blob/d74c96fd1c16b059d5d6d8b9cfca200b7ae3c40b/documentation/creating-swagger.md#NegativeResponses

**Comment by amarzavery** [comment]
This would not have happened, if there were examples of request and response for every operation. Recently we have updated the tooling that
- validates the examples against the definition specified in the swagger spec. This helps us catch model validation errors early on.
- The examples will be published to docs.microsoft.com website and customers can use these examples as a reference.

More info on x-ms-examples can be found [here](https://github.com/Azure/adx-documentation-pr/blob/master/tools/tools.md#swagger-model-validation).

@iouri-s - It would be great if DTL can also provide x-ms-examples for it's swagger spec.

**Comment by iouri-s** [comment]
@amarzavery - a spec defines most of the semantics of HTTP error codes and payloads for Azure Resource Manager resource providers. What is the point of each spec providing its own examples for validation (potentially violating the spec)?

**Comment by amarzavery** [comment]
We have caught increasing number of bugs that were not caught by sdk tests. Bugs like 
- properties present in the example but not defined in the swagger spec
- spec says that the property is a string but on the wire it is an integer
- response statuscodes defined in the example but the swagger spec does not have that defined or vice-versa.
- swagger spec says that the response has a schema (i.e. body is not empty) but the example (on the wire) says that actually there is no response body.
- the validation happens as a part of the CI, so we catch such issues early on.


Plus these examples serve as REST API documentation. Customers have complained for a long time that the REST API docs do not have good examples. The doc team is taking the swagger spec along with x-ms-examples and is generating docs that get published on docs.microsoft.com website.

**Comment by iouri-s** [comment]
@amarzavery - examples certainly provide value; however, they have nothing to do with the issue at hand. Even assuming there was an example that said that the 204 code was an error and was returning a payload, and even if the example was in strict conformance with the swagger spec and the service, that would have been wrong, because per the RP spec (and the HTTP protocol), 204 is not an error and must not have a payload. Specific issues about DTL spec are too removed from the topic at hand. Feel free to reach out on any appropriate channel and we can have a discussion.

**Comment by amarzavery** [comment]
The reason AutoRest generated client is treating "204" as an Error is because: "204" is not defined in the swagger spec. Any response status code not defined in the swagger for a given operation goes into the default bucket (which is error). 

_The model validator would have caught this and prompted the spec author to define 204 in the swagger. Then AutoRest generated client would not have thrown an error._

Hope that explains my point :).

As per [the swagger specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#responses-object)

### Responses Object
> A container for the expected responses of an operation. The container maps a HTTP response code to the expected response. It is not expected from the documentation to necessarily cover all possible HTTP response codes, since they may not be known in advance. However, it is expected from the documentation to cover a successful operation response and any known errors.

> The `default` can be used as the default response object for all HTTP codes that are not covered individually by the specification.

> The `Responses Object` **MUST** contain at least one response code, and it **SHOULD** be the response for a **successful operation call**.

**Comment by iouri-s** [comment]
Sorry, I do not get your point; the SDK team through some tool or code reviews for RP swagger specs can already flag these conditions, e.g. 204 not specified on delete or conversely, 404 or 400 explicitly declared (those should be handled by "default"). No examples are needed for those validations.

**Comment by amarzavery** [comment]
It is not mandatory that every service returns a 204 on delete. Irrespective of that, we will add a linter rule to flag a warning in the above scenario.

I am not saying that this cannot be done by any linter or code review tool. All I am saying is that with the type of tools we have today, this would have been caught by the model validator.

**Comment by iouri-s** [comment]
It is, in fact, mandatory that every ARM resource provider return 204 on delete. It may not be the case for non-ARM services, but if we got some consistency across ARM that would be a huge step forward.

From the RP spec:

Status Code
The resource provider can return 200 (OK) or 204 (NoContent) to indicate that the operation completed successfully. A 200 (OK) should be returned if the object exists and was deleted successfully; and a 204 (NoContent) should be used if the resource does not exist and the request is well formed. 
202 (Accepted) can be returned to indicate that the operation will complete asynchronously.
If the resource group does not exist, 404 (NotFound) will be returned by the proxy layer and will not reach the resource provider. 412 (PreconditionFailed) and other normal REST codes are acceptable as long as they match the REST guidelines.

**Comment by pomortaz** [comment]
In key vault service we don't return 204 for delete operation. Does 204 coming from the ARM proxy layer and every RP should accept this response? I actually haven't every seen a 204 response for a Key Vault operation.

**Comment by iouri-s** [comment]
Yes and yes. @pomortaz - Execute the same keyvault delete request twice and you will see that the second one returns a 204.

**Comment by iouri-s** [comment]
Also, if you do not return 204s for resources that do not exist, you should. Do not assume that the frontdoor won't call you.

**Comment by pomortaz** [comment]
@iouri-s thanks. I could reproduce the error. We'll get it fixed to include 204 as an accepted response.

**Comment by kirthik** [comment]
@pomortaz : Could you provide an ETA on when this would be fixed?

**Comment by pomortaz** [comment]
@kirthik can you please assign it to @RandalliLama. This is out of my scope now.
@RandalliLama can you please provide an ETA for this issue?
